Journal abbreviations dialog ported to JavaFX

boceckts opened the pull request:
I tried to extract the logic in a test driven approach to completely separate ui and ui-model and to have a high code test coverage. The design of the dialog was made after the ui-model was finisehd and is close to what was specified in #1396.

- [x] Change in CHANGELOG.md described
- [x] Tests created for changes
- [x] Screenshots added (for bigger UI changes)

![listview](https://cloud.githubusercontent.com/assets/15340757/17297484/38b74016-5806-11e6-94ff-4d42c61e584c.PNG)
![listvieweditname](https://cloud.githubusercontent.com/assets/15340757/17297485/39e412fc-5806-11e6-9ad4-bfeaa7aab699.png)
![listviewbuiltinlist](https://cloud.githubusercontent.com/assets/15340757/17297543/6f7a067e-5806-11e6-8f1c-e3979922db16.png)

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,315 @@
+
+                            @Override
+                            protected void updateItem(String s, boolean b) {
+                                super.updateItem(s, b);
```
Try using  a lambda for the callback, makes the code more readable:
http://stackoverflow.com/questions/23978676/convert-callback-into-java-8-lambda-expression

Please use more speaking  variable names for s and b, for example like here:
http://code.makery.ch/blog/javafx-8-tableview-cell-renderer/

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,282 @@
+        List<String> externalFiles = Globals.prefs.getStringList(JabRefPreferences.EXTERNAL_JOURNAL_LISTS);
+        externalFiles.forEach(name -> {
+            try {
+                openFile(new File(name));
```
Please use the nio Paths.get method

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,282 @@
+     * @param file to be opened
+     * @throws JabRefException if journal file with same path is already open
+     */
+    public void openFile(File file) throws JabRefException {
```
As mentioned above, please use nio methods

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,282 @@
+     */
+    public void openFile(File file) throws JabRefException {
+        AbbreviationsFile abbreviationsFile = new AbbreviationsFile(file);
+        if (journalFiles.contains(abbreviationsFile)) {
```
Why not simply use a Set?
No need for checking duplicates, as a Set implicitly ignores duplicates

boceckts replied:
I used the list as I wanted to have the elements in an order and I can't find a `SortedSetProperty`

tobiasdiez replied:
https://docs.oracle.com/javase/7/docs/api/java/util/LinkedHashSet.html ?

boceckts replied:
But there exists no Property for it which is needed in order to use the bindings.

Siedlerchr replied:
You could try the `SortedList `
Here is a quick example I found:
http://code.makery.ch/blog/javafx-8-tableview-sorting-filtering/

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,282 @@
+     */
+    public void addAbbreviation() throws JabRefException {
+        Abbreviation abbreviation = new Abbreviation(abbreviationsName.get(), abbreviationsAbbreviation.get());
+        if (abbreviations.contains(abbreviation)) {
```
see above, try using a set

Siedlerchr commented:
Looks good :+1:  ! 
Instead of the "+" Button  would suggest simply the text "Add".
What is the purpose of the "-" Icon? Deletion? If yes, a kind of trash can symbol would be more suitable

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -1,39 +1,43 @@
-    public AbbreviationsFile(File file) {
-        super(file.getAbsolutePath());
+    public AbbreviationsFile(String filePath) {
+        path = FileSystems.getDefault().getPath(filePath.replaceAll("\\\\", "/"));
```
You could simply use `Paths.get(filePath)` this is a shorthand for the FileSystems.getDefault...

What are you trying to do here? I do not understand the Replace stuff...  In case you are trying to escape sth like Slashes, that is absolutely not necessary, as the Path methods are able to handle both relative and absolute.
Have a look here for some examples
https://docs.oracle.com/javase/tutorial/essential/io/pathOps.html

boceckts replied:
The idea was indeed to escape the backslashes for windows systems...but thanks for the tip

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -43,4 +47,35 @@ public void WriteOrCreate() throws FileNotFoundException, IOException {
+    @Override
+    public boolean equals(Object obj) {
+        if (obj instanceof AbbreviationsFile) {
+            return this.path.equals(((AbbreviationsFile) obj).getPath());
```
Simpler: `Objects.equals(Object,Object) `

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -43,4 +47,35 @@ public void WriteOrCreate() throws FileNotFoundException, IOException {
+
+    @Override
+    public int hashCode() {
+        return new HashCodeBuilder().append(path).toHashCode();
```
Simpler:
`return Objects.hash(...)`

boceckts commented:
I updated the description and added screenshots for the latest commit. Is it really the best idea to add the "+" to the end of the list of journal abbrevaitions? This would mean the user always has to scroll to the bottom to add more journal abbreviations to the list. In my opinion this is less ideal than having a button that is always visible and in the same place. Any thoughts?

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,308 @@
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
```
Could you add for all classes, that you've added, a short comment?

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,308 @@
+        AbbreviationsFile ieeeFile = new AbbreviationsFile(ieeeList, "IEEE built in list");
+        journalFiles.add(ieeeFile);
+    }
+
```
please add short comment

tschechlovdev commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,554 @@
+        savelyAddNewFileToViewModel("newTestFile.txt");
+
+        // size of the list of journal files should be incremented by one
+        Assert.assertEquals(size + 1, viewModel.journalFilesProperty().size());
```
could you just use a explicity value instead of size +1?

tschechlovdev commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,554 @@
+                + "MoreEntries = ME" + Globals.NEWLINE + "Entry = E" + Globals.NEWLINE + "";
+        String actual = Files.contentOf(new File(testFile4Entries), StandardCharsets.UTF_8);
+        Assert.assertEquals(expected, actual);
+
```
could you in general try to split up the test methods into more test methods?

tobiasdiez commented:
Personally, I prefer inline editing for lists/tables with only a handful of columns. See for example http://designingwebinterfaces.com/ultimate-guide-to-table-ui-patterns. The very first row could be an "insert row", i.e. it shows (in italic) "Create new abbreviation" and when you click you just have an empty textfield to insert the new abbreviation. 

Other from that, I really like how the dialog looks. Maybe change the close/remove icon? Btw what is the purpose of this "JabRef Lists" button at the bottom?

boceckts commented:
@tobiasdiez so when clicking on the edit icon on the right the column should be editable?
The "JabRef Lists" button shows the jabref built in journal abbreviations which are not modifyable. In the swing version there is a "view" button to show these abbreviations in a separate window.

tobiasdiez commented:
@boceckts Yes exactly, clicking edit (and I would also support double clicking) makes the row editable. Mhh on second thought, one might not even need the edit button but show it directly in the cell upon mouse hover and a single click then makes the cell editable. Decide based on what is easier to implement. 

For the "JabRef List" button, you also have a JabRef built in list in the drop-down. What is the difference here?

boceckts commented:
@tobiasdiez The built in lists in the drop down will only appear when the toggle button is activated

tobiasdiez commented:
Is there a need to hide it from the drop-down list? If not, then I would just remove the toggle button.

boceckts commented:
@tobiasdiez I just thought it's nice to give the user  the option to only display their custom lists

koppor commented:
@boceckts Please remove "JabRef lists" button. :sunflower:

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,546 @@
+            @Override
+            protected void updateItem(Boolean isPseudoAbbreviation, boolean isEmpty) {
+                super.updateItem(isPseudoAbbreviation, isEmpty);
+                if (isPseudoAbbreviation != null) {
```
is this really needed ? When can this be null?

boceckts replied:
Yes it is necessary because we have to call `setGraphic(null)` if the item to update is null. Otherwise we get some weird graphic bugs when switching from a custom list to a built in list since the graphics don't get reset.

![journalabb](https://cloud.githubusercontent.com/assets/15340757/17476618/31371e1e-5d61-11e6-92d3-621bf8192548.PNG)

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,118 @@
+ * This class provides a model for abbreviation files.
+ * It actually doesn't save the files as objects but rather saves
+ * their paths. This also allows to specify pseudo files as placeholder objects.
+ *
```
remove this unnecessary empty lines in the comments.

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,546 @@
+            protected void updateItem(Boolean isPseudoAbbreviation, boolean isEmpty) {
+                super.updateItem(isPseudoAbbreviation, isEmpty);
+                if (isPseudoAbbreviation != null) {
+
```
remove empty line

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,546 @@
+
+                    if (!isEmpty) {
+                        if (isEditableAndRemovable.get()) {
+                            if (!isPseudoAbbreviation) {
```
why don't put the if-statements together with and? Then you have only one time ``else setGraphic(null)``

boceckts replied:
But I only need to call `setGraphic(null)` on two of the four if statements. However I can join the two if statements with no else block.

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,342 @@
+    /**
+     * Method to change the currentAbbrevaition property to a new abbreviation.
+     * The name and the actual abbreviation will be taken from the abbreviationsName
+     * and abbreviationsAbbreviation properties.
```
add an empty line here.

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,342 @@
+            throw new JabRefException("Name can not be empty");
+        } else if (abbreviation.trim().isEmpty()) {
+            throw new JabRefException("Abbrevaition can not be empty");
+        } else {
```
else shouldn't be necessary here

tschechlovdev commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,342 @@
+                file.WriteOrCreate();
+            } catch (FileNotFoundException e) {
+                logger.debug(e.getLocalizedMessage());
+            } catch (IOException ioe) {
```
you can put the catch blocks together ``FileNotFoundException | IOException e``

boceckts replied:
I found out `FileNotFoundException` is even redundant if I catch the `IOException` first.

simonharrer commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,614 @@
+            editAbbreviation(testAbbreviation);
+        } catch (JabRefException e1) {
+            Assert.fail("The same abbreviation already exists in the current list.");
+        }
```
do not catch exception in tests, just throw them. In this case, you hide the exception from JUnit, and the user does not get the stack trace, only your short text.

tobiasdiez commented on CHANGELOG.md:
```diff
@@ -30,6 +30,7 @@ to [sourceforge feature requests](https://sourceforge.net/p/jabref/features/) by
 - Integrity check table can be sorted by clicking on column headings
 - Add tab which shows the MathSciNet review website if the `MRNumber` field is present.
 - Partly switched to new UI technology (JavaFX). Redesigned key bindings dialog.
+- Partly switched to new UI technology (JavaFX). Redesigned journal abbreviations dialog.
```
Make the different redesigned dialogs a subbulled of "Partly switched to ..."

boceckts replied:
I now moved them to subbullets and to ## JavaFX Changes

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationViewModel.java:
```diff
@@ -0,0 +1,100 @@
+        this.name.set(name);
+    }
+
+    public void setAbbreviation(String name) {
```
rename parameter "name"

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+ * It actually doesn't save the files as objects but rather saves
+ * their paths. This also allows to specify pseudo files as placeholder objects.
+ */
+public class AbbreviationsFile {
```
add suffix "ViewModel"

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+    public AbbreviationsFile(List<AbbreviationViewModel> abbreviations, String name) {
+        this.abbreviations.addAll(abbreviations);
+        this.name = name;
+        this.path = null;
```
Make path an Optional<Path>

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+    }
+
+    public void readAbbreviations() throws FileNotFoundException {
+        List<Abbreviation> abbreviationList = JournalAbbreviationLoader.readJournalListFromFile(path.toFile());
```
A NPE is thrown if this file is a built-in list.

boceckts replied:
Now caught by `if (path.isPresent())` and if false, an appropriate `FileNotFoundException` is thrown

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+
+    private final SimpleListProperty<AbbreviationViewModel> abbreviations = new SimpleListProperty<>(
+            FXCollections.observableArrayList());
+    private final ReadOnlyBooleanProperty isPseudoFile;
```
Maybe rename isPseudoFile -> isBuiltinList to better reflect what the actual purpose of this field is.

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+     *
+     * @throws IOException
+     */
+    public void WriteOrCreate() throws IOException {
```
This logic contains code which I would declare as "logic", namely how abbreviations are serialized. Please extract the code to a new method `writeAbbreviations(path, encoding?, List<Abbreviation>). This new method may life in `Abbrevation` or a new class `AbbrevationWriter`.

boceckts replied:
Extracted to new class AbbreviationWriter in logic package

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFile.java:
```diff
@@ -0,0 +1,132 @@
+    public void WriteOrCreate() throws IOException {
+        if (!isPseudoFile.get()) {
+            try (OutputStream outStream = Files.newOutputStream(path);
+                    OutputStreamWriter writer = new OutputStreamWriter(outStream, Globals.prefs.getDefaultEncoding())) {
```
Do we really want the user decide about the encoding? I would propose to always use UTF-8 if this works.

boceckts replied:
Changed to `StandardCharset.UTF_8`

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+    }
+
+    private void setButtonStyles() {
+        journalFilesBox.setPromptText(Localization.lang("No abbreviation files loaded"));
```
This styling can't be done in FXML?

boceckts replied:
Hm I must have missed that somehow, yes it can (and now will) be done by fxml

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+        setUpTable();
+        setBindings();
+        setButtonStyles();
+        new Thread(new Runnable() {
```
JavaFX seems to have a new interface for background tasks: docs.oracle.com/javafx/2/threads/jfxpub-threads.htm
To be honest, I have no idea about its advantages/disadvantages. Could you pls do some research and maybe switch the code to `Task<...>`

boceckts replied:
`Task` seems like a similar class as `Runnable`, while `Service` is the JavaFX equivalent to a `Thread` it seems. The JavaFX classes support continuous feedback of the state of the background task and work with data binding to show a progress bar for the current status. `Service` can only be run from the JavaFX Application `Thread`. Unlike Threads they can also be restarted multiple times and you can add event handlers if the execution was successfull/stopped/unsuccessfull/etc.

In this particular scenario it shouldn't matter though. For the moment I changed it to a `Task` running in a `Thread` but `Thread` and `Service` are easily interchangable if needed.

tobiasdiez replied:
Thanks for doing the research. Can you please add the first paragraph to the Wiki?

boceckts replied:
Should I add it as a new chapter in the Code HowTos as something like "Working with Threads"?

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+                loadingLabel.setVisible(true);
+                progressIndicator.setVisible(true);
+                viewModel.createFileObjects();
+                journalFilesBox.getSelectionModel().selectLast();
```
In the best case the viewModel also takes care of which item is selected (as usual through binding). Can you change this if it is not too hard.

boceckts replied:
I extracted the methode to the viewmodel, the currently selected file is already bound.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+    private void setUpTable() {
+        journalAbbreviationsTable.setOnKeyPressed(event -> {
+            if ((event.getCode() == KeyCode.DELETE) && isEditableAndRemovable.get()) {
+                if ((viewModel.currentAbbreviationProperty().get() != null)
```
combine ifs

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+public class ManageJournalAbbreviationsView extends FXMLView {
+
+    private final ManageJournalAbbreviationsViewModel viewModel = new ManageJournalAbbreviationsViewModel();
+    private final BooleanProperty isEditableAndRemovable = new SimpleBooleanProperty(true);
```
Why is `isEditableAndRemovable` not part of the viewModel?

boceckts replied:
Good point I will try to move it

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+
+    private void setUpTable() {
+        journalAbbreviationsTable.setOnKeyPressed(event -> {
+            if ((event.getCode() == KeyCode.DELETE) && isEditableAndRemovable.get()) {
```
Just notify the viewModel about the pressed key and let it decide what to do, i.e.
`if (event.getCode == KeyCode.DELETE) viewModel.deleteCurrentAbbreviation()` and the rest of the logic is in the viewModel.

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+        journalTableAbbreviationColumn.setCellFactory(cell -> new JournalAbbreviationsAbbreviationTableEditingCell());
+        journalTableEditColumn.setCellValueFactory(cellData -> cellData.getValue().isPseudoAbbreviationProperty());
+        journalTableDeleteColumn.setCellValueFactory(cellData -> cellData.getValue().isPseudoAbbreviationProperty());
+        journalTableEditColumn.setCellFactory(column -> new TableCell<AbbreviationViewModel, Boolean>() {
```
I really don't like these complicated cell factories...stupid javafx (not your fault and nothing to do 😺 )

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+                if (isPseudoAbbreviation != null) {
+                    if (!isEmpty) {
+                        if (isEditableAndRemovable.get()) {
+                            if (!isPseudoAbbreviation) {
```
Remove negation from if statement and switch if and else blocks

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+                                Text graphic = new Text(IconTheme.JabRefIcon.EDIT.getCode());
+                                graphic.getStyleClass().add("icon");
+                                setGraphic(graphic);
+                                setOnMouseClicked(evt -> {
```
the braces are not necessary for one-line statements, i.e. `setOnMouseClicked(evt -> editAbbrevation)` is sufficient.

boceckts replied:
done in all cases

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+                                    editAbbreviation();
+                                });
+                            } else {
+                                Text graphic = new Text(IconTheme.JabRefIcon.ADD.getCode());
```
Is it easy to change the icon class a bit so that you can write
`setGraphic(IconTheme.JabRefIcon.ADD.getGraphic())`?

boceckts replied:
Created a new method in `JabRefIcon` class

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+        viewModel.currentAbbreviationProperty().addListener((observable, oldvalue, newvalue) -> {
+            journalAbbreviationsTable.getSelectionModel().select(newvalue);
+        });
+        journalAbbreviationsTable.getSelectionModel().selectedItemProperty()
```
a bidirectional binding between `journalAbbreviationsTable.getSelectionModel().selectedItemProperty()` and `viewModel.currentFileProperty()` does not work?

boceckts replied:
I will look into this

boceckts replied:
`journalAbbreviationsTable.getSelectionModel().selectedItemProperty()` is read only and can therefore not be bound. The `journalAbbreviationsTable.itemsProperty()` is a `ObjectProperty<ObservableList<AbbreviationViewModel>>` which unfortunately can not be bound to my `SimpleListProperty<AbbreviationsFileViewModel>.

tobiasdiez replied:
According to http://stackoverflow.com/questions/32349660/how-to-bidirectionally-bind-to-a-property-of-an-object-set-to-the-selected-item there should be a way to bind it using https://github.com/TomasMikula/EasyBind. This library seems really nice and thus I'm in favor of integrating it (if the stackoverflow answer is correct and it indeed works).

boceckts replied:
Ignore my comment above, the lists can actually be bound without an extra library as I just found out when playing around with the EasyBind Library... :expressionless:
But I don't think `journalAbbreviationsTable.getSelectionModel().selectedItemProperty().addListener(...`  and similar can be replaced.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+            if (newvalue != null) {
+                isEditableAndRemovable.set(!newvalue.isPseudoFileProperty().get());
+            }
+            removeJournalAbbreviationsButton.setDisable((newvalue == null) || !isEditableAndRemovable.get());
```
I would prefer if the disable property is bound to something in the viewmodel

boceckts replied:
I think the disable property is readonly for ui elements, therefore it can not be bound to other properties.

boceckts replied:
Just realised there are two properties, disbaledProperty is read only the other one can be bound

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+    @FXML
+    private void addNewFile() {
+        FileChooser chooser = new FileChooser();
+        chooser.getExtensionFilters().add(new ExtensionFilter("TXT files (*.txt)", "*.txt"));
```
Localize "files", i.e. Localization.lang("%0 file", "TXT") + "(*.txt)"

boceckts replied:
Changed it, but the console says:
>  could not get message translation for "%0 file" for locale en

tobiasdiez replied:
Should work after master is merged into JavaFX.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+
+    private void selectNewAbbreviation() {
+        int lastRow = viewModel.abbreviationsCountProperty().get() - 1;
+        journalAbbreviationsTable.scrollTo(lastRow);
```
Again, in the best case all of this is handled by binding to the viewModel. (if it is not too difficult)

boceckts replied:
The selection is already bound to the view model, unfortunately the table view doesn't automatically scroll to the selected row...maybe the `scrollTo()` call can be moved to the property listener of the current abbreviation property to always scroll if the current abbreviation changes.

boceckts replied:
Moving the call of `scrollTo()` to the current abbreviation property listener means the view will always scroll so the selected row is the first row in the visible view of rows. This means if if I select a row in the middle of the view it will still scroll so the selected row is the first row. Therefore I would rather leave it like this.

tobiasdiez replied:
ok that is fine with me

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+
+    @FXML
+    private void addAbbreviation() {
+        viewModel.abbreviationsNameProperty().set("Name");
```
Why do you set name and abbreviation here?

tobiasdiez replied:
Can these two set-lines also be moved to `viewModel.addAbbreviation`? Also localize the strings.

boceckts replied:
Yes I will change it tomorrow, the test also have to be adjusted then.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+            try {
+                viewModel.openFile(file.getAbsolutePath());
+                journalFilesBox.getSelectionModel().selectLast();
+            } catch (JabRefException e) {
```
I don't like that a generic `JabRefException` is thrown.

boceckts replied:
I introduced three new Exception extending JabRefException in net.sf.jabref.logic.journals:
`DuplicatedJournalAbbreviationException`, `DuplicatedJournalFileException` and `EmptyFieldException`

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+
+    @FXML
+    private void saveAbbreviations() {
+        viewModel.saveExternalFilesList();
```
combine this into viewModel.saveEverthing() (or something like that) ?

boceckts replied:
changed it

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+            textField.focusedProperty().addListener(new ChangeListener<Boolean>() {
+
+                @Override
+                public void changed(ObservableValue<? extends Boolean> arg0, Boolean arg1, Boolean arg2) {
```
better names for arguments

boceckts replied:
done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,543 @@
+            textField.focusedProperty().addListener(new ChangeListener<Boolean>() {
+
+                @Override
+                public void changed(ObservableValue<? extends Boolean> arg0, Boolean arg1, Boolean arg2) {
```
here too

boceckts replied:
also done

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,339 @@
+ * abbreviations and files. To visualize the model one can bind the properties
+ * to ui elements.
+ */
+public class ManageJournalAbbreviationsViewModel {
```
I stop here for now... (just a reminder for myself)

Siedlerchr commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsView.java:
```diff
@@ -0,0 +1,527 @@
+    @FXML
+    private void addNewFile() {
+        FileChooser chooser = new FileChooser();
+        chooser.getExtensionFilters().add(new ExtensionFilter(Localization.lang("%0 file", "TXT"), "*.txt"));
```
Have a look at the new FileExtensions Enum I created in #1336. Already inlcudes the Loalization.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFileViewModel.java:
```diff
@@ -0,0 +1,136 @@
+    public void writeOrCreate() throws IOException {
+        if (!isBuiltInList.get()) {
+            List<Abbreviation> actualAbbreviations = Lists.newArrayList();
+            abbreviations.forEach(abb -> {
```
I think the stream interface is more readable here, i.e. something along the lines `abbreviations.stream().filter( !isPseudoAbbreviation).map( getAbbreviationObject).collect()`.

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/AbbreviationsFileViewModel.java:
```diff
@@ -0,0 +1,136 @@
+    }
+
+    public String getAbsolutePath() {
+        return path.get().toAbsolutePath().toString();
```
Be careful. The path may be an empty optional. I would propose to change the return value to Optional<String> and write `return path.map(toAbsolutePath().toString());`

tobiasdiez commented on src/main/resources/l10n/JabRef_da.properties:
```diff
@@ -1667,6 +1667,7 @@ Message=
 
 MathSciNet_Review=
 Reset_Bindings=
+<<<<<<< acdb36f664504aed9b5db5a173094ba075ee6c18
```
Remove this merge marker

tobiasdiez commented on src/main/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsViewModel.java:
```diff
@@ -0,0 +1,356 @@
+                extFiles.add(file.getAbsolutePath());
+            }
+        });
+        Globals.prefs.putStringList(JabRefPreferences.EXTERNAL_JOURNAL_LISTS, extFiles);
```
We are trying to reduce the occurrences of `Globals.prefs`. Thus I would like to see that the view model is independent on `Globals.prefs`. At least inject a JabRefPrefences class as a dependency and work on this class. Probably better it is to just work on an injected `JournalAbbreviationPreferences`. @oscargus does it make sense to introduce a field in `JournalAbbreviationPreferences` corresponding to `EXTERNAL_JOURNAL_LISTS`?

tobiasdiez commented on src/main/java/net/sf/jabref/logic/journals/Abbreviation.java:
```diff
@@ -1,29 +1,66 @@
     public Abbreviation(String name, String abbreviation) {
-        this.name = java.util.Objects.requireNonNull(name).trim();
-        this.abbreviation = java.util.Objects.requireNonNull(abbreviation).trim();
+        this.name.set(java.util.Objects.requireNonNull(name).trim());
```
Remove the `java.util` part by using the proper import.

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+            try (OutputStream outputStream = new FileOutputStream(testFile)) {
+                outputStream.write(content.getBytes(StandardCharsets.UTF_8));
+            }
+        } catch (IOException e) {
```
Don't catch this exception here. If something goes wrong with the initialization, the tests should better fail.

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+        viewModel.editAbbreviation();
+    }
+
+    private void saveFilesToPreferences() throws DuplicatedJournalFileException {
```
probably a better name: `addFilesToViewModel`

boceckts replied:
But the important thing here is that the files are getting saved to the preferences..maybe I change it to `addFourTestFileToViewModelAndPreferences`.

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+    }
+
+    @Test
+    public void testAddFile() throws DuplicatedJournalFileException {
```
please rename the tests so that they reflect what you are testing. here `addFileIncreasesCounterOfOpenFiles`

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+
+    @Test
+    public void testAddFile() throws DuplicatedJournalFileException {
+        int size = viewModel.journalFilesProperty().size();
```
You are starting with an empty viewModel and so size == 0, right? In this case just remove the size variable.

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+        selectLastJournalFile();
+
+        // size of the list of journal files should be incremented by one
+        Assert.assertEquals(size, viewModel.journalFilesProperty().size());
```
size is always 1, so use hard-coded value.

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+
+        // size of the list of journal files should be incremented by one
+        Assert.assertEquals(size, viewModel.journalFilesProperty().size());
+        // our test file has 3 abbreviations
```
4?

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+        viewModel.changeActiveFile(viewModel.journalFilesProperty().get(1));
+
+        // size of the list of journal files should be incremented by two
+        Assert.assertEquals(size, viewModel.journalFilesProperty().size());
```
again remove size variable

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+                .observableArrayList(Globals.journalAbbreviationLoader.getBuiltInAbbreviations());
+        ObservableList<AbbreviationViewModel> actual = viewModel.abbreviationsProperty().get();
+        ObservableList<Abbreviation> actualAbbreviations = FXCollections.observableArrayList();
+        actual.forEach(abbViewModel -> actualAbbreviations.add(abbViewModel.getAbbreviationObject()));
```
use map function, `viewModel.getAbbrevations().stream().map(getAbbrevaitonObject).collect()`

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+                .observableArrayList(Globals.journalAbbreviationLoader.getOfficialIEEEAbbreviations());
+        ObservableList<AbbreviationViewModel> actual = viewModel.abbreviationsProperty().get();
+        ObservableList<Abbreviation> actualAbbreviations = FXCollections.observableArrayList();
+        actual.forEach(abbViewModel -> actualAbbreviations.add(abbViewModel.getAbbreviationObject()));
```
see above

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+
+        // simulate open file button
+        viewModel.addNewFile(emptyTestFile);
+        size++;
```
again size variable, remove it everywhere pls

tobiasdiez commented on src/test/java/net/sf/jabref/gui/journals/ManageJournalAbbreviationsTest.java:
```diff
@@ -0,0 +1,560 @@
+        selectLastJournalFile();
+        Abbreviation testAbbreviation = new Abbreviation("YetAnotherEntry", "YAE");
+        addAbbrevaition(testAbbreviation);
+        editAbbreviation(testAbbreviation);
```
no assertion?

tobiasdiez commented:
I just made some final remarks. After these are fixed, this PR can be merged in my opinion. @koppor pls do the final review & merge, thanks.

@boceckts (and the all others who worked on this PR): good job! I really like your code, especially that you wrote tests to implement the logic! One can see that you learned a lot in the last few weeks.

boceckts commented:
I think I now addressed all of your comments.

tobiasdiez commented:
I'll merge this now, since it looks good to me and all the javafx code will be finally reviewed anyway.

koppor commented:
Does not work in `javafx`. @boceckts Can you please check?

```
Exception in thread "JavaFX Application Thread" java.lang.IllegalStateException: Cannot load net.sf.jabref.gui.journals.managejournalabbreviations
        at com.airhacks.afterburner.views.FXMLView.loadSynchronously(FXMLView.java:94)
        at com.airhacks.afterburner.views.FXMLView.initializeFXMLLoader(FXMLView.java:101)
        at com.airhacks.afterburner.views.FXMLView.getView(FXMLView.java:113)
        at net.sf.jabref.gui.journals.ManageJournalAbbreviationsView.showAndWait(ManageJournalAbbreviationsView.java:201)
        at net.sf.jabref.gui.journals.ManageJournalsAction.lambda$actionPerformed$0(ManageJournalsAction.java:21)
        at com.sun.javafx.application.PlatformImpl.lambda$null$173(PlatformImpl.java:295)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$174(PlatformImpl.java:294)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$null$148(WinApplication.java:191)
        at java.lang.Thread.run(Thread.java:745)
Caused by: javafx.fxml.LoadException: No resources specified.
/C:/git-repositories/jabref/jabref/build/resources/main/net/sf/jabref/gui/journals/managejournalabbreviations.fxml:22

        at javafx.fxml.FXMLLoader.constructLoadException(FXMLLoader.java:2597)
```

boceckts commented:
@koppor I will look into it, did this happen after the javafx branch has been updated?

boceckts commented:
@koppor Ok so I tracked the issue and came to the conclusion that the this pr #1918 causes the exception. As described in the pr's description @tobiasdiez had to out comment the line `bundle = Localization.getMessages();` since otherwise the localization tests would fail. After adding it back it works for me again.

tobiasdiez commented:
@boceckts do the localization tests also work after readding the `bundle = ...` line?

boceckts commented:
No but it gives me the same errors when the line is out commented, I'm not sure if this statement is causing the failing tests..

koppor commented:
Yes, this happens with the current javafx branch (merge of master). I did not check the state before the merge. My gut feeling is that the merge broke it. But I'm not sure.

I is **not** about failing tests, it is about that "Options -> Manage Journal Abbreviations" is a noop.

boceckts commented:
Yes it's about opening the dialog. But still, re-adding the line as described by me two comments ago fixes the issue for me.

tobiasdiez commented:
Reenabled in #1935.
